src/libserialize/leb128.rs

```diff
@@ -1,46 +1,14 @@
-#[inline]
-pub fn write_to_vec(vec: &mut Vec<u8>, byte: u8) {
-    vec.push(byte);
-}
-
-#[cfg(target_pointer_width = "32")]
-const USIZE_LEB128_SIZE: usize = 5;
-#[cfg(target_pointer_width = "64")]
-const USIZE_LEB128_SIZE: usize = 10;
-
-macro_rules! leb128_size {
-    (u16) => {
-        3
-    };
-    (u32) => {
-        5
-    };
-    (u64) => {
-        10
-    };
-    (u128) => {
-        19
-    };
-    (usize) => {
-        USIZE_LEB128_SIZE
-    };
-}
-
 macro_rules! impl_write_unsigned_leb128 {
     ($fn_name:ident, $int_ty:ident) => {
         #[inline]
         pub fn $fn_name(out: &mut Vec<u8>, mut value: $int_ty) {
-            for _ in 0..leb128_size!($int_ty) {
-                let mut byte = (value & 0x7F) as u8;
-                value >>= 7;
-                if value != 0 {
-                    byte |= 0x80;
-                }
-
-                write_to_vec(out, byte);
-
-                if value == 0 {
+            loop {
```

Comment:
Could optimize this further by reserving enough space for the largest possible `value` outside the loop, then use `unsafe` to do unchecked writes. `set_len` could be avoided until after the loop.

Should avoid another branch in each iteration (unless LLVM already hoists `reserve` out).

Reply:
I tried this:
```rust
        pub fn $fn_name(out: &mut Vec<u8>, mut value: $int_ty) {
            let mut len = out.len();
            out.reserve(19); // Maximum possible length = ceiling(128/7)
            loop {
                if value < 0x80 {
                    unsafe { *out.get_unchecked_mut(len) = value as u8 };
                    len += 1;
                    break;
                } else {
                    unsafe { *out.get_unchecked_mut(len) = ((value & 0x7f) | 0x80) as u8 };
                    len += 1;
                    value >>= 7;
                }
            }
            unsafe {
                out.set_len(len);
            }
        }
```
It was a measurable slowdown. In practice, single-byte writes are easily the most common (more than 50%) and the average length is less than 2, which might partly explain the outcome.